### DISCUSSION TITLE
richness with sobs

### DISCUSSION THREAD

**Discussion by iBuitron** [scikit-bio/scikit-bio#2122] [General]
```
import numpy as np
import pandas as pd
from skbio.diversity import alpha_diversity

```
# Create a sample DataFrame
```
data = {
    "Sample_1": [10, 7, 5, 0],  # Count of individuals per species in the first sample
    "Sample_2": [3, 0, 1, 0],  # Count of individuals per species in the second sample
    "Sample_3": [0, 2, 0, 4],  # Count of individuals per species in the third sample
}

species = ["Festuca orthophylla", "Poa aequigluma", "Lachemilla pinnata", "aaa"]

df = pd.DataFrame(data, index=species)

print(df)
                     Sample_1  Sample_2  Sample_3
Festuca orthophylla        10         3         0
Poa aequigluma              7         0         2
Lachemilla pinnata          5         1         0
aaa                         0         0         4

```

# Calculate alpha diversity using our DataFrame
```
counts_matrix = df.T.to_numpy()  # Transpose to get samples as rows
results = alpha_diversity(metric="sobs", counts=counts_matrix, ids=df.columns)

print(results)
Sample_1    3
Sample_2    2
Sample_3    2
```

And:

in Just one case:

```
results = alpha_diversity(metric="sobs", counts=data["Sample_2"])
print(results)

2 species
```

My question is:
There is a case on my table that i get a specie with a 0 population (but is becaus indirect observation, such as feces) 
how can i keep this specie with 0 population in the richness result? 

Oh the case of Sample_2, it will be 4 species instead 2.

Context: The table comes from a species record with a single abundance column, if the species is present (either because individuals were found or it was through indirect observation) this is placed in the abundance column as a value >0 or 0 (if it is indirect observation).

Shall i use a custom func like?:
```
def sobs_including_zeros(counts):
    return len(counts)
```
    
o there a skitbio way?

**Comment by qiyunzhu** [comment]
Hello @iBuitron Thanks for your interest. The alpha diversity metrics in scikit-bio remove all zero cells prior to the calculation. If you intend to include zeros, you may do `len(counts)` as you suggested, or you can add a pseudocount to eliminate zeros before applying the scikit-bio function, for example: `alpha_diversity(metric="sobs", counts=data["Sample_2"] + 1)`. Hope it helps!
